osquery/database/plugins/rocksdb.cpp

```diff
@@ -114,6 +120,41 @@ class RocksDBDatabasePlugin : public DatabasePlugin {
 /// Backing-storage provider for osquery internal/core.
 REGISTER_INTERNAL(RocksDBDatabasePlugin, "database", "rocksdb");
 
+class RocksDBGuard {
+ public:
+  RocksDBGuard() : my_tid_(platformGetTid()), locked_(false) {}
+
+  bool safeEnter() {
+    if (!owner_tid_is_valid_ || owner_tid_ != my_tid_) {
+      mu_.lock();
+      owner_tid_ = my_tid_;
+      owner_tid_is_valid_ = true;
+      locked_ = true;
+    }
+    return locked_;
+  }
+
+  ~RocksDBGuard() {
+    if (locked_) {
+      owner_tid_is_valid_ = false;
+      mu_.unlock();
+    }
+  }
+
+ private:
+  int my_tid_;
+  bool locked_;
+
+  static std::mutex mu_;
+  static int owner_tid_;
+  // linux has no notion of an invalid thread id, hence owner_tid_is_valid_
+  static bool owner_tid_is_valid_;
+};
+
+std::mutex RocksDBGuard::mu_;
+int RocksDBGuard::owner_tid_;
```

Comment:
I believe this is the correct way to implement this. pthread does not provide for a reserved value that can be used to denote a distinct invalid pthread id, so initializing owner_tid_ to some value, even 0, is incorrect. The recommended way to do this is to use a separate boolean (in our case, owner_tid_is_valid_), to indicate whether the value in owner_tid_ is a valid thread_id or not. 

Here is a discussion I thought was interesting: http://stackoverflow.com/questions/16599463/what-is-the-value-range-of-thread-and-process-id